tests/unit/Snak/ExtendedPropertyValueSnakTest.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+namespace Wikibase\DataModel\Tests\Snak;
+
+use DataValues\StringValue;
+use Wikibase\DataModel\Entity\PropertyId;
+use Wikibase\DataModel\Snak\ExtendedPropertyValueSnak;
+
+/**
+ * @covers Wikibase\DataModel\Snak\ExtendedPropertyValueSnak
+ *
+ * @group Wikibase
+ * @group WikibaseDataModel
+ * @group WikibaseSnak
+ *
+ * @licence GNU GPL v2+
+ * @author Adam Shorland
+ */
+class ExtendedPropertyValueSnakTest extends SnakObjectTest {
+
+	public function constructorProvider() {
+		return array(
+			'No extras' => array(
+				true,
+				new PropertyId( 'P1' ),
+				new StringValue( 'a' ),
+				array(),
+			),
+			'2 extras' => array(
+				true,
+				new PropertyId( 'P9001' ),
+				new StringValue( 'bc' ),
+				array( 'foo' => new StringValue( 'foo' ), 'bar' => new StringValue( 'bar' ) ),
+			),
+			'fail - Integer key' => array(
+				'InvalidArgumentException',
+				new PropertyId( 'P9001' ),
+				new StringValue( 'bc' ),
+				array( new StringValue( 'foo' ) ),
+			),
+			'fail - not a value' => array(
+				'InvalidArgumentException',
+				new PropertyId( 'P9001' ),
+				new StringValue( 'bc' ),
+				array( 'foo' => 'bar' ),
+			),
+		);
+	}
+
+	public function getClass() {
+		return 'Wikibase\DataModel\Snak\ExtendedPropertyValueSnak';
+	}
+
+	/**
+	 * This test is a safeguard to make sure hashes are not changed unintentionally.
+	 */
+	public function testHashStability() {
+		$snak = new ExtendedPropertyValueSnak(
+			new PropertyId( 'P1' ),
+			new StringValue( 'a' ),
+			array( 'foo' => new StringValue( 'foo' ) )
+		);
+		$hash = $snak->getHash();
+
+		// @codingStandardsIgnoreStart
+		$expected = sha1( 'C:49:"Wikibase\DataModel\Snak\ExtendedPropertyValueSnak":110:{a:3:{i:0;i:1;i:1;C:22:"DataValues\StringValue":1:{a}i:2;a:1:{s:3:"foo";C:22:"DataValues\StringValue":3:{foo}}}}' );
```

Comment:
I suggest to tackle that together with the other snaks. Which will be a breaking change anyway. Since this blocks getting rid of the whole pile of old serialization code in Wikibase Lib, let's not block the introduction of this class on such a change.